### PR TITLE
[Dubbo-3886] Integrate with SOFARegistry.

### PR DESCRIPTION
**Pull request by ujjboy**
## What is the purpose of the change

#3874 

- Add Registry extension of [SOFARegistry](https://github.com/alipay/sofa-registry)
- Add test case.

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [x] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [x] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by ujjboy** [comment]
@ralf0131  Please review this PR.

**Comment by ralf0131** [comment]
I will try to do it today or tomorrow..

File: dubbo-all/pom.xml [hunk 1]
```diff
@@ -254,6 +254,20 @@
             <scope>compile</scope>
             <optional>true</optional>
         </dependency>
+        <dependency>
+            <groupId>org.apache.dubbo</groupId>
+            <artifactId>dubbo-registry-nacos</artifactId>
```
Comment (ralf0131) [inline dubbo-all/pom.xml]: This is unnecessary to have. Please sync with the latest branch.

File: dubbo-all/pom.xml [hunk 2]
```diff
@@ -530,6 +544,8 @@
                                     <include>org.apache.dubbo:dubbo-registry-redis</include>
                                     <include>org.apache.dubbo:dubbo-registry-consul</include>
                                     <include>org.apache.dubbo:dubbo-registry-etcd3</include>
+                                    <include>org.apache.dubbo:dubbo-registry-nacos</include>
```
Comment (ralf0131) [inline dubbo-all/pom.xml]: So does this.

File: dubbo-registry/dubbo-registry-sofa/pom.xml [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+        <dependency>
+            <groupId>org.apache.dubbo</groupId>
+            <artifactId>dubbo-common</artifactId>
+            <version>${project.version}</version>
+            <optional>true</optional>
+        </dependency>
+
+        <dependency>
+            <groupId>com.alipay.sofa</groupId>
```
Comment (ralf0131) [inline dubbo-registry/dubbo-registry-sofa/pom.xml]: External dependency should go to dubbo-dependency-pom/pom.xml

File: dubbo-registry/dubbo-registry-sofa/pom.xml [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        <dependency>
+            <groupId>org.apache.dubbo</groupId>
+            <artifactId>dubbo-config-api</artifactId>
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+
+        <!-- test modules -->
+        <dependency>
+            <groupId>com.alipay.sofa</groupId>
```
Comment (ralf0131) [inline dubbo-registry/dubbo-registry-sofa/pom.xml]: So does this.

File: dubbo-registry/dubbo-registry-sofa/src/main/java/org/apache/dubbo/registry/sofa/SofaRegistry.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+import static org.apache.dubbo.registry.sofa.SofaRegistryConstants.ADDRESS_WAIT_TIME_KEY;
+import static org.apache.dubbo.registry.sofa.SofaRegistryConstants.DEFAULT_GROUP;
+import static org.apache.dubbo.registry.sofa.SofaRegistryConstants.LOCAL_DATA_CENTER;
+import static org.apache.dubbo.registry.sofa.SofaRegistryConstants.LOCAL_REGION;
+
+/**
+ * The Sofa registry.
+ *
+ * @author <a href="mailto:ujjboy@qq.com">GengZhang</a>
```
Comment (ralf0131) [inline dubbo-registry/dubbo-registry-sofa/src/main/java/org/apache/dubbo/registry/sofa/SofaRegistry.java]: Author name should be removed.

File: dubbo-registry/dubbo-registry-sofa/src/main/java/org/apache/dubbo/registry/sofa/SofaRegistryConstants.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.dubbo.registry.sofa;
+
+/**
+ * @author <a href="mailto:ujjboy@qq.com">GengZhang</a>
```
Comment (ralf0131) [inline dubbo-registry/dubbo-registry-sofa/src/main/java/org/apache/dubbo/registry/sofa/SofaRegistryConstants.java]: Author name should be removed.

File: dubbo-rpc/dubbo-rpc-thrift/pom.xml [hunk 7]
```diff
@@ -69,5 +69,10 @@
             <version>${project.parent.version}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (ralf0131) [inline dubbo-rpc/dubbo-rpc-thrift/pom.xml]: Why this change is needed?

File: dubbo-dependencies-bom/pom.xml [hunk 8]
```diff
@@ -140,8 +140,9 @@
         <hessian_lite_version>3.2.5</hessian_lite_version>
         <swagger_version>1.5.19</swagger_version>
         <spring_test_version>4.3.16.RELEASE</spring_test_version>
-
+        <junit_version>4.12</junit_version>
```
Comment (lovepoem) [inline dubbo-dependencies-bom/pom.xml]: Now the we use junit5 for unit case

File: dubbo-dependencies-bom/pom.xml [hunk 9]
```diff
@@ -511,8 +511,26 @@
...
+            </dependency>
+            <dependency>
+                <groupId>com.alipay.sofa</groupId>
+                <artifactId>registry-test</artifactId>
+                <version>${sofa_registry_version}</version>
+                <scope>test</scope>
+            </dependency>
 
             <!-- Test lib -->
+            <dependency>
```
Comment (lovepoem) [inline dubbo-dependencies-bom/pom.xml]: Now the we use junit5 for unit case, please remove it

File: dubbo-registry/dubbo-registry-sofa/src/test/java/org/apache/dubbo/registry/sofa/SofaRegistryTest.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+import com.alipay.sofa.registry.server.test.TestRegistryMain;
+import org.apache.dubbo.common.logger.Logger;
+import org.apache.dubbo.common.logger.LoggerFactory;
+import org.apache.dubbo.config.ApplicationConfig;
+import org.apache.dubbo.config.ProtocolConfig;
+import org.apache.dubbo.config.ReferenceConfig;
+import org.apache.dubbo.config.RegistryConfig;
+import org.apache.dubbo.config.ServiceConfig;
+import org.junit.AfterClass;
```
Comment (lovepoem) [inline dubbo-registry/dubbo-registry-sofa/src/test/java/org/apache/dubbo/registry/sofa/SofaRegistryTest.java]: Please use junit5

File: dubbo-config/dubbo-config-spring/pom.xml [hunk 11]
```diff
@@ -120,7 +120,6 @@
         <dependency>
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
-            <version>4.12</version>
```
Comment (lovepoem) [inline dubbo-config/dubbo-config-spring/pom.xml]: Now spring module still use `<version>4.12</version>` , it's a special usage . Please retain it

**Comment by ujjboy** [comment]
@ralf0131 @lovepoem  Hi, I have changed the code according to review, but CI only build error in `openjdk11`, Can you retry it?

**Comment by ralf0131** [comment]
@lovepoem What is you opinion on this?

**Comment by ralf0131** [comment]
I checked the license of dubbo-registry-sofa dependencies. All looks good.

**Comment by lovepoem** [comment]
lgtm
